### PR TITLE
Add aten mkldnn conv2d backward operator

### PR DESCRIPTION
**Pull request by XiaobingSuper**
### mkldnn backward ops list:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20567) Add aten mkldnn conv2d backward operator :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20570) Add aten mkldnn backward ops: relu, linear and reshape :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20571) Add aten mkldnn backward ops: max_pool2d, avg_pool2d and adaptive_avg_poo2d :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20572) Add aten mkldnn batchnorm backward operator :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20573) Add aten mkldnn zero_ operator💚 
- [ ] \(https://github.com/pytorch/pytorch/pull/20575) Add mkldnn mul operator 💚 

Enable mkldnn backward which can improve the traning performance about 2x for mode resnext101.

### CODE REVIEW

**Comment by mingfeima** [comment]
@bddppq Hi, the backward integration is done and please review the code.
Feel free to invite more reviewers.

**Comment by XiaobingSuper** [comment]
@uyongw,  @Jianhui-Li, @jgong5.

**Review by dzhulgakov — CHANGES_REQUESTED** [review]
In general looks ok. Two questions:
* CI is failing
* Did you benchmark potential perf change from introducing ideep in the backward path? it should be ok, but best to double check

**Comment by mingfeima** [comment]
@dzhulgakov so what is the proposed system config for training? double sockets or single socket?
@XiaobingSuper post benchmark numbers here.

However, we have a known issue here that memory buffer is not cached (output tensors need to be allocated everytime), this is going to add significant overhead for the training scenario.

Customized memory allocator is a solution but seems to be against the current design. Suggestions?

**Comment by jgong5** [comment]
> However, we have a known issue here that memory buffer is not cached (output tensors need to be allocated everytime), this is going to add significant overhead for the training scenario.
> 
> Customized memory allocator is a solution but seems to be against the current design. Suggestions?

@dzhulgakov The performance issue @mingfeima pointed out is caused by `clear_page` overhead on large buffer allocation with `malloc`. This happens when PyTorch allocates large activation buffer which is a typical case during training since the batch size is usually large. As you know, whether `clear_page` is triggered is controlled by M_MMAP_THRESHOLD (http://man7.org/linux/man-pages/man3/mallopt.3.html) which default to 128KB and on 64-bit system, the upper bound is 32MB which is not a big number w.r.t. typical training workloads. I notice that PyTorch has caching allocator in the GPU path and probably CPU can use similar idea to avoid calling `malloc` again and again for output tensors. Currently, we have a quick implementation of caching allocator inside ideep and @XiaobingSuper will share the training performance numbers with and without caching allocator FYI.

**Comment by XiaobingSuper** [comment]
Runing this [benchmark](https://github.com/XiaobingSuper/convnet-benchmark-py) on SKX-6148 with 2 sockets, we can get **2X** performance improvement using mkldnn compared to native path:

models | native path(img/s) | mkldnn path(img/s) | speed up
-- | -- | -- | --
resnet50 | 28.27 | 67.81 | 239.87%
resnext101 | 8.73 | 19.71 | 225.77%

 The following is the details logs:

- Native path:

```
### using OMP_NUM_THREADS=40
### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1

Running on device:  Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz
Running on torch: 1.2.0a0+69e3229
Running on torchvision: 0.3.0a0+8837e0e

ModelType: resnet50, Kernels: nn Input shape: 128x3x224x224
nn                              :forward:    1803.12 (ms)      70.99 (imgs/s)
nn                             :backward:    2725.13 (ms)
nn                               :update:       4.72 (ms)
nn                                :total:    4528.25 (ms)      28.27 (imgs/s)
ModelType: resnext101_32x8d, Kernels: nn Input shape: 128x3x224x224
nn                              :forward:    6042.85 (ms)      21.18 (imgs/s)
nn                             :backward:    8624.28 (ms)
nn                               :update:      13.07 (ms)
nn                                :total:   14667.13 (ms)       8.73 (imgs/s)
```

- MKLDNN path:
```
### using OMP_NUM_THREADS=40
### using KMP_AFFINITY=granularity=fine,compact,1,0
### using KMP_BLOCKTIME=1

Running on device:  Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz
Running on torch: 1.2.0a0+69e3229
Running on torchvision: 0.3.0a0+8837e0e

ModelType: resnet50, Kernels: nn Input shape: 128x3x224x224
nn                              :forward:     659.40 (ms)     194.12 (imgs/s)
nn                             :backward:    1228.16 (ms)
nn                               :update:       4.69 (ms)
nn                                :total:    1887.55 (ms)      67.81 (imgs/s)
ModelType: resnext101_32x8d, Kernels: nn Input shape: 128x3x224x224
nn                              :forward:    2259.24 (ms)      56.66 (imgs/s)
nn                             :backward:    4234.59 (ms)
nn                               :update:      13.00 (ms)
nn                                :total:    6493.83 (ms)      19.71 (imgs/s)
```

**Comment by XiaobingSuper** [comment]
Adding the performance using [ idepp caching allocator](https://github.com/intel/ideep/blob/df4b7ff8ea6a3b12cb5150ca8eaddfbe3b9f842b/include/ideep/allocators.hpp#L88) on SKX-6148 with 2 sockets, there has **1.36x** performance improvement at least for inference with large batch size and has **1.32x** performance improvement at least for training.

1. For inference using MKLDNN with large batch size:

models | MKLDNN_default allocator (img/s) | MKLDNN_caching allocator(img/s) | speed up
-- | -- | -- | --
resnet50 | 196.91 | 289.69 | 147.12%
resnext101 | 58.23 | 79.77 | 136.99%

2. For training using MKLDNN:

models | MKLDNN_default allocator (img/s) | MKLDNN_caching allocator(img/s) | speed up
-- | -- | -- | --
resnet50 | 67.81 | 96.05 | 141.65%
resnext101 | 19.71 | 26.2 | 132.93%

**Comment by dzhulgakov** [comment]
Adding CPU caching allocator is something we also discussed. The problem is mostly alleviated by using better malloc - e.g. using jemalloc (http://jemalloc.net/jemalloc.3.html). Unfortunately, I don't think there's a safe way to package it with prebuilt pytorch binaries - it has to be preloaded for the entire Python process.

Overall, this PR looks pretty good to me, also cc @zheng-xq to take a look

**Comment by XiaobingSuper** [comment]
@zheng-xq , can you help review it?

**Comment by XiaobingSuper** [comment]
@bddppq, can you help review this code, perharps we can first merge this PR which can unify the mkldnn convolution code. Thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, I rebase the code again, can you help review it when you have free time? Thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, the test failed cases are not related to this PR. Thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by Jianhui-Li** [comment]
@dzhulgakov regarding the jemalloc.  We tried out Jemalloc and TCmalloc, but both creates extra dependencies and doesn’t support NUMA. TF expose NUMA as subdevice and user can do NUMA-aware memory allocation.  The memory allocator makes big difference when running trainning or offline inference with large batch size, since malloc may trigger large overhead since the malloc of large chunk memory causes clear page overhead. We observed +30% benefit on throughput. Are there any plan for Pytorch to support similar capability, if not, do you think it would work for us to implement one and PR it?

**Comment by XiaobingSuper** [comment]
This PR reopened in https://github.com/pytorch/pytorch/pull/36121.
